### PR TITLE
Windows: TerminateOnLastHandleClosed

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard <jhoward@microsoft.com>

No idea how no-one had noticed this before, but it's been like this
forever. Verified that running a container and killing the daemon
does now kill the container. (Note this is in the v1 schema path,
and already correct in the v2 schema runtime when using containerd.)

@jterry75 FYI.

### CODE REVIEW

**Review by thaJeztah — APPROVED** [review]
SGTM

**Comment by olljanat** [comment]
@jhowardmsft good catch. This explains some weird conditions which we have seen on our environment and which why we actually have restart script which first stop docker service before it actually restart server. You can find more details about it from https://github.com/microsoft/hcsshim/issues/437#issuecomment-455313994 / MS case 119010419510056 if you want to verify that this actually fixes also that issue.

However I suggest that we merge #39240 first and then you can update this integration test to working with Windows too.
https://github.com/moby/moby/blob/b958b430aa2bfe72ab93ae2eac72fb73c324c5c8/integration/container/restart_test.go#L16-L114

It is also interesting that currently live restore is not supported by Windows containers but it sounds that it would be actually quite simple to implement.

**Comment by jhowardmsft** [comment]
Live restore is not supported. It will be by containerd. From a brief look at that test and from the way in which Windows is different to Linux, I'm not sure that test will be that easy to replicate. Regardless, the fix here is pretty essential independent of the test.

**Review by jterry75 — APPROVED** [review]
LGTM (not a maintainer)

**Comment by arsing** [comment]
>No idea how no-one had noticed this before, but it's been like this
forever.

As I mentioned to @jhowardmsft over email, a build of 79b5b19439e19cd367f3a78511ce8c15102659d8 (released as Moby 3.0.5) does reliably stop the container when the daemon exits. So the change that broke it is relatively recent, somewhere in https://github.com/moby/moby/compare/79b5b19439e19cd367f3a78511ce8c15102659d8...04ff4a2ba42f47e967ed006baa97f4e146dba1c6 I did not try to bisect it.

>Verified that running a container and killing the daemon
does now kill the container.

In my testing, the container still remains running when `dockerd.exe` is killed and restarted, even with this PR applied. The entrypoint process of the container does exit, but that happens even without this PR.

**Comment by olljanat** [comment]
@arsing did you also check that this is not related to Windows updates?

Your explanation on https://github.com/Azure/iotedge/pull/1314#issuecomment-502232613 actually makes me thinking that is this actually same issue why #39320 was needed?

What comes to possible root cause my best guess is #39209 so if you have env where you can easily test this it would be good idea to try revert it.

**Comment by arsing** [comment]
@olljanat

>@arsing did you also check that this is not related to Windows updates?

What specifically do you want me to check?

If you mean the Windows version, all my testing for both Moby 3.0.5 and master was on RS5 (10.0.17763) with latest updates as of that day.

>Your explanation on https://github.com/Azure/iotedge/pull/1314 (comment) actually makes me thinking that is this actually same issue why #39320 was needed?

I don't know. When I started the email conversation with jhowardmsft , he told me *this* PR would fix the issue in https://github.com/Azure/iotedge/pull/1314 as well. There's a bunch of things - closing the last handle, detaching the sandbox VHD, the container remaining "Up" even though its entrypoint process already exited - and I don't know what their interplay is.

>What comes to possible root cause my best guess is #39209 so if you have env where you can easily test this it would be good idea to try revert it.

I'll look into that, though I'm more concerned about the container remaining "Up" rather than the VHD remaining attached.

---

FWIW this is the summary of my testing that I'd sent to jhowardmsft:

>With dockerd.3.0.5.exe (my build of Moby 3.0.5 release, ie 79b5b194) or dockerd.3.0.5.39594.exe (my build of 3.0.5 with PR 39594 cherry-picked):
>
>- Container shows as "Exited" (per `docker ps -a`) after dockerd.exe is killed and restarted.
>- After restarting dockerd.exe, `docker start` fails to start the container with "hcsshim::ActivateLayer - failed failed in Win32: The process cannot access the file because it is being used by another process. (0x20)".
>- `docker rm -f` succeeds. Alternatively, it works to run `DetachVirtualHardDisk()` and then `docker start`.
>
>
>With dockerd.master.exe (my build of master, ie 04ff4a2b) or dockerd.master.39594.exe (my build of master with PR 39594 merged):
>
>- Container shows as "Up" (as per `docker ps -a`) after dockerd.exe is killed and restarted, though cmd.exe process inside container does stop running.
>- After restarting dockerd.exe, `docker rm -f` blocks indefinitely. `docker exec` and `docker top` fail with "no such container".
>- `gwmi MsVm_MountedStorageImage` still shows the sandbox.vhdx. Calling `DetachVirtualHardDisk()` on it succeeds but does not change anything about the `docker rm -f`, etc results.
>
>
>So the PR does not appear to change anything about the respective builds that it's merged into, and also the behavior of master is much more broken than the 3.0.5 release.

(The container used for this testing is just a regular `mcr.microsoft.com/windows/nanoserver:1809-amd64`)

**Comment by jhowardmsft** [comment]
@arsing There are multiple issues in your specific scenario. The fix here is absolutely correct and an omission. I can repro the other case (of not being able to start a container after your scenario). I am seeing if a mitigation can be made in docker, or whether a platform fix may be required. The two things are orthogonal though.

**Comment by arsing** [comment]
Right, I don't mean that this PR shouldn't be merged at all. Auto-closing the handle is of course a good thing, and it being orthogonal to whatever's causing the issue in my scenario would explain why this PR doesn't make a difference there.

I only commented here because you initially told me this PR *would* resolve my scenario. Please let me know when you find something about that.

---

Apart from that, maybe:

>No idea how no-one had noticed this before, but it's been like this
>forever. Verified that running a container and killing the daemon
>does now kill the container.

... in the commit message should be changed to something else? Since as I said it's a) a new regression after Moby 3.0.5 rather than "forever" despite the handle not being auto-closed in either case, and b) killing the daemon does not kill the container, as you found. But that's your call.

**Comment by jhowardmsft** [comment]
It resolves part of your scenario. Let's try and resolve everything in one go. I'll temporarily close this PR while I'm working with others here to root cause some other oddities. (Of course, your scenario is brute-force killing the daemon which is never a good idea ;) )

**Comment by thaJeztah** [comment]
@jterry75 @jhowardmsft @taylorb-microsoft is someone working on the "other oddities"? Wondering if it would _hurt_ to take this patch as-is (and work on the other parts as a follow up)

The issue is also mentioned in https://freddysblog.com/2019/08/07/zombie-containers-skip-this-build/

**Comment by thaJeztah** [comment]
^ fix for the above was found; https://github.com/moby/moby/pull/39703

**Comment by krytarowski** [comment]
What does really happen when TerminateOnLastHandleClosed is false? The container is empty but alive? Is there any use-case for this functionality to not terminate by default on the last handle close?

**Comment by jterry75** [comment]
@krytarowski - Its for reattach. If you wanted to be able to shut down a given process and then attach to a pre-existing container. With TerminateOnLastHandleClosed == true when the holding process shuts down the container is forcibly terminated.

**Pull request closed without merging**
